node/src/components/consensus/highway_core/active_validator/round_success_meter.rs

```diff
@@ -0,0 +1,100 @@
+            latest_handled_round_index: None,
+        }
+    }
+
+    pub fn change_exponent(&mut self, new_exp: u8, timestamp: Timestamp) {
+        self.rounds = VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER);
+        self.round_exp = new_exp;
+        self.round_length = 1u64 << new_exp;
+        self.last_exponent_change_index = (timestamp.millis() / self.round_length) as usize;
+        self.latest_handled_round_index = None;
+    }
+
+    /// Registers the success or failure of finalizing a block within a round, and returns the new
+    /// round exponent based on whether we are keeping up with the blocks or not.
+    pub fn handle_finalized_block(
+        &mut self,
+        block_timestamp: Timestamp,
+        time_finalized: Timestamp,
+    ) -> u8 {
+        // in case the block timestamp isn't at the beginning of the round
+        let block_round_start_index = (block_timestamp.millis() / self.round_length) as usize;
+        let finalization_round_index = (time_finalized.millis() / self.round_length) as usize;
+        let round_successful = finalization_round_index == block_round_start_index;
+        // if we finalized a block at round N, and the last round we handled is N-K, we mark all
+        // rounds between rounds N-K+1 and N-1 as failed
+        let num_rounds_to_fail = self.latest_handled_round_index.map_or(
+            block_round_start_index - self.last_exponent_change_index,
+            |latest| block_round_start_index.saturating_sub(latest + 1),
+        );
+        for _ in 0..num_rounds_to_fail {
+            self.rounds.push_front(false);
+        }
+        // mark round N as successful
+        // this will be None if latest_handled_round_index was None, or block round index was
+        // greater than the latest round index
+        let round_index_to_mark = self
+            .latest_handled_round_index
+            .and_then(|latest| latest.checked_sub(block_round_start_index));
+        match round_index_to_mark {
+            None => {
+                self.rounds.push_front(round_successful);
+                self.latest_handled_round_index = Some(block_round_start_index);
+            }
+            Some(index) if round_successful => {
+                if let Some(success_ref) = self.rounds.get_mut(index) {
+                    *success_ref = true;
+                }
+            }
+            _ => (),
+        }
+        self.clean_old_rounds();
+        self.new_exponent(block_round_start_index)
+    }
+
+    fn clean_old_rounds(&mut self) {
+        while self.rounds.len() > NUM_ROUNDS_TO_CONSIDER {
+            self.rounds.pop_back();
+        }
+    }
+
+    fn count_failures(&self) -> usize {
+        self.rounds.iter().filter(|&success| !success).count()
+    }
+
+    fn new_exponent(&self, block_round_start_index: usize) -> u8 {
+        if block_round_start_index % 2 == 0 && self.count_failures() > MAX_FAILED_ROUNDS {
```

Comment:
Should that actually be `!= 0`?
Because if the block we just got was in an _even_ round index, the next one will be odd, so we can _not_ increase the exponent?

But maybe we can ignore that condition entirely, because `ActiveValidator::round_exp` already checks whether the change is allowed anyway? Incrementing `next_round_exp` one round too early shouldn't be a problem.